open-api/rest-catalog-open-api.yaml

```diff
@@ -537,6 +537,113 @@ paths:
         5XX:
           $ref: '#/components/responses/ServerErrorResponse'
 
+  /v1/{prefix}/namespaces/{namespace}/tables/{table}/preplan:
```

Comment:
More fundamental question: Since we're going down a route of adding a REST catalog defining which capabilities are supported, does every server which supports planning also need to support preplanning? Or are these different capabilities? @nastra @rdblue 

IMO I think these should be separate capabilities. Pre-planning seems like it would require a lot more state to maintain on the server and it wouldn't make sense to me to lump the 2 capabilities because then the barrier for a REST catalog to technically support planning is significantly higher.

Reply:
We can discuss this, I didn't mean to imply that's the only way! There may be legitimate reasons why it should be a single capability. Is there any particular advantages you see with having a single capability?

Reply:
> does every server which supports planning also need to support preplanning

I think no, a server can decide to only implement one API. The server can **require** the client to go through the preplan API by indicating a server capability (e.g. preplanning), and in that case if the client still directly hits the plan endpoint without a plan task, then a 421 Misdirected Request should be thrown.

Reply:
> > does every server which supports planning also need to support preplanning
> 
> I think no, a server can decide to only implement one API. The server can **require** the client to go through the preplan API by indicating a server capability (e.g. preplanning), and in that case if the client still directly hits the plan endpoint without a plan task, then a 421 Misdirected Request should be thrown.

I agree with this. So to confirm we are saying that these two should be separate capabilities, `preplanning` and `planning`, however only the `plan` api is required to be implemented where as preplanning is not required.
@jackye1995 @amogh-jahagirdar

Reply:
what do you mean "required to be implemented"? Server can also choose to not implement that, and it will fallback to client-side planning

Reply:
@jackye1995 Yes techincally server does not to have implement any of these, but I thought the discussion/thread started with `which capabilities are supported by server`, if server does not choose to implement `plan` then there are no capabilities essentially as `preplan` response is just passed as input to `plan`.

Reply:
@jackye1995 Are we going to make edits to server capabilities for scan planning in this pr or is this something that will be down in another pr?

 And did we align for `scan-planning` capability that a server can choose to implement none of these endpoints, can implement both `preplan` and `plan`, or just implement `plan`? Im not sure if there is a use case where a server just only implements `preplan` by itself?

Reply:
I've also been thinking about how to handle this, and I think what is more important is how define the default behavior for plan/preplan.

The `plan` endpoint is always required because that's the only way to get the list of `FileScanTask` to process. The `preplan` endpoint can be very simple -- even as simple as always returning the exact same opaque plan task for all requests. In other words, it would make no sense for `preplan` to return 501 because it is just as easy to fake a plan task and have the `plan` endpoint do all the real work.

My conclusion is that it isn't worth a separate capability for preplanning and everything should support it.

That's why I think the more important choice is whether `plan` or `preplan` is called first. For default behavior, there are two scenarios:
1. If the client calls `plan` without calling `preplan`
2. If the client calls `preplan` and then `plan`

In scenario 1, the client calls `plan` and the service can decide whether to redirect it to `preplan` using 521. If the service doesn't use `preplan` or the query is small then it returns file scan tasks to the client in a single request (fastest). If the service does want to use `preplan` for a large scan then it can quickly return 521 to redirect the client to `preplan`. That `preplan` fallback case takes at least 3 requests (521 from `plan`, `preplan`, and `plan` with a plan task) but it is likely that there are more requests because of the size of the scan.

In scenario 2, the client always calls `preplan` and is sent plan tasks, then calls `plan` one or more times. There are always at least 2 requests but there are probably more for large scans.

I would choose scenario 1 for the default because in the average case I think it will perform better. I think most of the time the service will return the file scan tasks in a single request because small dimension-like tables are the most common. And when there is 1 extra request to plan a large scan, performance will be dominated by metadata scanning time anyway.

If we do go with a default to call the `plan` endpoint first, then that is more of a reason not to add a separate capability for `preplan` because the `plan` endpoint can choose to never redirect the client.

Reply:
I'm not sure if we really should force the client into certain directions. For example, I think PyIceberg will mainly go to `plan` directly as it is more likely to consume smaller tables than your typical ETL workloads.

> I think no, a server can decide to only implement one API. The server can require the client to go through the preplan API by indicating a server capability (e.g. preplanning), and in that case if the client still directly hits the plan endpoint without a plan task, then a 421 Misdirected Request should be thrown.

So then a redirect would be sufficient? 

> And did we align for scan-planning capability that a server can choose to implement none of these endpoints, can implement both preplan and plan, or just implement plan?

I think there are situations, like Amogh described, where you don't want to accumulate too much state on the server-side (for example, data sharing servers), so in that case I would expect that it would only expose the plan endpoint. But in such a case, the pre-plan could also be a no-op like Ryan suggested.

Reply:
> In other words, it would make no sense for preplan to return 501 because it is just as easy to fake a plan task and have the plan endpoint do all the real work.

This is a good point, I wasn't really considering it from this angle but yeah it's not much of a lift to just return an opaque plan task where the actual plan endpoint does the work. this is one example in favor of not having to introduce another capability. 

> In scenario 1, the client calls plan and the service can decide whether to redirect it to preplan using 521.

@rdblue I think you mean 422? https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422 but agree with the idea

Overall the approach of working from how plan/preplan will be called first or not seems reasonable, at least it establishes what a sane default would be for the average case (calling plan table first). I'm not sure I'd entirely base the decision on capabilities based off of this since as @Fokko said different clients can do different things depending on their use case but the first point around "it's as easy just to return an opaque token as it is to throw a 501" convinced me that it's not worth having a separate capability. 

I think that's actually a good heuristic for determining if something should be a new capability; is the lift heavy to have a simple implementation compared to throwing a 501 or not? In this case it's essentially a no-op implementation so the complexity of having a different capability probably isn't worth it.

>I'm not sure if we really should force the client into certain directions.

@Fokko so I actually don't know if we're forcing clients, more so establishing what a sane default is for which endpoint should be called first. This default would be established off the average case, which I think I agree with @rdblue that most cases plan endpoint would just work without redirection.

 If a client has to preplan in it's average case workloads then I think they can still do that, maybe there's a client configuration to enable that but that's really in the implementation I'd say.

Reply:
I think I did mean either 421 or 422 and not 521. The return codes that redirect the client to the preplan endpoint.

I think there's a fair point that if we only have one capability, clients can call either endpoint first. I'd definitely _recommend_ calling `plan` first, but that's up to the client and, more importantly, because we don't _need_ to require a certain behavior we should not require a certain behavior. Specs should be flexible and leave room for implementations to be smart.

Reply:
> I think there's a fair point that if we only have one capability, clients can call either endpoint first. I'd definitely recommend calling plan first, but that's up to the client and, more importantly, because we don't need to require a certain behavior we should not require a certain behavior. Specs should be flexible and leave room for implementations to be smart.

I agree with this that we should not require a client to have to call `preplan` or `plan` first. However `plan` to me is the core capability and that the client will likely be calling this first, especially when the client does not need to leverage a `preplan` to distribute this work up if for a small table scan.

If we say this is the recommendation but not requirement where are we mentioning this? Is this within the `plan` endpoint description?  @rdblue @amogh-jahagirdar @Fokko

Reply:
With the current structure of these endpoints, this is no longer needed. I'll resolve it.